cmd/otel-allocator/allocation/least_weighted/allocator.go

```diff
@@ -0,0 +1,210 @@
+		Name: "opentelemetry_allocator_collectors_allocatable",
+		Help: "Number of collectors the allocator is able to allocate to.",
+	})
+	timeToAssign = promauto.NewHistogramVec(prometheus.HistogramOpts{
+		Name: "opentelemetry_allocator_time_to_allocate",
+		Help: "The time it takes to allocate",
+	}, []string{"method"})
+	_ strategy.Allocator = &LeastWeightedAllocator{}
+)
+
+/*
+	Load balancer will serve on an HTTP server exposing /jobs/<job_id>/targets
+	The targets are allocated using the least connection method
+	Load balancer will need information about the collectors in order to set the URLs
+	Keep a Map of what each collector currently holds and update it based on new scrape target updates
+*/
+
+// LeastWeightedAllocator makes decisions to distribute work among
+// a number of OpenTelemetry collectors based on the number of targets.
+// Users need to call SetTargets when they have new targets in their
+// clusters and call SetCollectors when the collectors have changed.
+type LeastWeightedAllocator struct {
+	// m protects collectors and targetItems for concurrent use.
+	m     sync.RWMutex
+	state strategy.State
+
+	log logr.Logger
+}
+
+// TargetItems returns a shallow copy of the targetItems map.
+func (allocator *LeastWeightedAllocator) TargetItems() map[string]*strategy.TargetItem {
+	allocator.m.RLock()
+	defer allocator.m.RUnlock()
+	targetItemsCopy := make(map[string]*strategy.TargetItem)
+	for k, v := range allocator.state.TargetItems() {
+		targetItemsCopy[k] = v
+	}
+	return targetItemsCopy
+}
+
+// Collectors returns a shallow copy of the collectors map.
+func (allocator *LeastWeightedAllocator) Collectors() map[string]*strategy.Collector {
+	allocator.m.RLock()
+	defer allocator.m.RUnlock()
+	collectorsCopy := make(map[string]*strategy.Collector)
+	for k, v := range allocator.state.Collectors() {
+		collectorsCopy[k] = v
+	}
+	return collectorsCopy
+}
+
+// findNextCollector finds the next collector with fewer number of targets.
+// This method is called from within SetTargets and SetCollectors, whose caller
+// acquires the needed lock.
+func (allocator *LeastWeightedAllocator) findNextCollector() *strategy.Collector {
+	var col *strategy.Collector
+	for _, v := range allocator.state.Collectors() {
+		// If the initial collector is empty, set the initial collector to the first element of map
+		if col == nil {
+			col = v
+		} else {
+			if v.NumTargets < col.NumTargets {
+				col = v
+			}
+		}
+	}
+	return col
+}
+
+// addTargetToTargetItems assigns a target to the next available collector and adds it to the allocator's targetItems
+// This method is called from within SetTargets and SetCollectors, whose caller acquires the needed lock.
+// This is only called after the collectors are cleared or when a new target has been found in the tempTargetMap
+func (allocator *LeastWeightedAllocator) addTargetToTargetItems(target *strategy.TargetItem) {
+	chosenCollector := allocator.findNextCollector()
```

Comment:
There is a possibility that the target allocator starts up without any collectors and the collector returned here could be nil.  I don't think we have to fix it in this pr, but I'm inclined to make an issue about better handling of the case of 0 collectors.

Reply:
Definitely agreed, this was a problem before and is a case we need to handle more effectively. I don't believe we are able to access this code block with 0 collectors however, because we short circuit in the `HandleCollectors()` method